lib/assets/javascripts/_modules/search.js

```diff
@@ -0,0 +1,171 @@
+      downloadSearchIndex();
+      attach();
+    };
+
+    function downloadSearchIndex() {
+      updateTitle('Loading search index')
+      $.ajax({
+        url: '/search.json',
+        cache: true,
+        method: 'GET',
+        success: function(data) {
+          lunrData = data;
+          lunrIndex = lunr.Index.load(lunrData.index);
+          $(document).trigger('lunrIndexLoaded');
+        }
+      });
+    }
+
+    function attach() {
+      // Search functionality on search text input
+      $searchInput.on('input', function (e) {
+        e.preventDefault();
+        var query = $(this).val();
+        search(query);
+      });
+
+      // Set focus on the first search result instead of submiting the search  
+      // form to Google
+      $searchForm.on('submit', function(e) {
+        e.preventDefault();
+        $searchResults.find('.search-result__title a').first().focus();
+      });
+
+      // Closing the search results, move focus back to the search input
+      $searchResultsClose.on('click', function(e) {
+        e.preventDefault();
+        $searchInput.focus();
+        hideResults();
+      });
+    }
+
+    function searchIndex(query) {
+      var results = [];
+      lunrIndex.search(query).forEach( function (item, index) {
+        if ( index < maxSearchEntries ) {
+          results.push(lunrData.docs[item.ref]);
+        }
+      });
+
+      return results;
+    }
+
+    function search(query) {
+      if(query === '') {
+        hideResults();
+        return;
+      }
+      showResults();
+      // The index has not been downloaded yet, exit early and wait.
+      if(!lunrIndex) {
+        $(document).on('lunrIndexLoaded', function() {
+          search(query);
+        });
+        return;
+      }
+      results = searchIndex(query);
+      renderResults(query);
+      updateTitle();
+    }
+
+    function renderResults(query) {
+      var output = '';
+      if (results.length == 0) {
+        output += '<p>Nothing found</p>';
+      }
+      output += '<ul>';
+      for(var index in results) {
+        var result = results[index];
+        var content = processContent(result.content, query);
+        output += '<li class="search-result">';
+        output += '<h3 class="search-result__title">';
+        output += '<a href="' + result.url + '">';
+        output += result.title;
+        output += '</a>';
+        output += '</h3>';
+        if(typeof content !== 'undefined') {
+          output += '<p>' + content + '</p>';          
+        }
+        output += '</li>';
+      }
+      output += '</ul>';        
+
+      $searchResults.html( output );
+    }
+
+    function processContent(content, query) {
+      var output;
+      content = '<div>'+  content + '</div>'; 
+      content = $(content).mark(query);
+
+      // Split content by sentence.
+      var sentences = content.html().replace(/(\.+|\:|\!|\?|\r|\n)(\"*|\'*|\)*|}*|]*)/gm, "|").split("|");
```

Comment:
Can you create tests for this code that generates the snippets?

I think I understand what this is doing, but it's not super clear what the intended behaviour is, and if we end up changing that regex (for example) I wouldn't want to break edge cases that you've already considered.

I expect we may want to truncate these further since matching sentences could be quite long, but I'd rather wait and see if it's actually an issue.

Reply:
> I expect we may want to truncate these further since matching sentences could be quite long, but I'd rather wait and see if it's actually an issue.

Yeah I can see that being an issue. It's going to be tricky to truncate without accidentally removing required context. Let's see what comes up.